src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs

```diff
@@ -394,7 +395,7 @@ private async Task SetStateAsync(NetworkPeerState newState)
                 if ((newState == NetworkPeerState.Failed) || (newState == NetworkPeerState.Offline))
                 {
                     this.logger.LogDebug("Communication with the peer has been closed.");
-
+                    this.asyncProvider.Signals.Publish(new PeerDisconnected(this.Inbound, this.PeerEndPoint, this.DisconnectReason?.Reason, this.DisconnectReason?.Exception));
```

Comment:
I'm wondering whether we should perhaps only do something like this in debug mode? What do you think?

Reply:
These events aren't strictly related to debug (except probably the Message sent/received message that contains the payload) so if we really want to bound these to something, probably should be to the -diagpeerstats argument (or better, the equivalent variable that holds its value because diagnostic can be even enabled by swagger)

I thought already about having them wrapped around an `if` condition when I created this feature but these events can be used by future features like SignalR to relay those events to a GUI to reflect changes on connected peers, etc... and so requiring these use cases to enable debug or diagpeerstats (that indeed add some work, even if negligible because it uses an async loop to handle these events) 


The overlay is really negligible (just instantiate a lightweight class and call a method that will do nothing except checking in a dictionary, because if no one need this events, nothing else will happen) and I don't expect we have thousands of these events per seconds (not even tens in this case because disconnections happens very few times)

We may reconsider this if we see performance issues (I don't expect them tbh) and I'll make use of diagpeerstats to wrap at least messages events in that case, we may even use pooling to prevent to instantiate new messages if needed (but will complicate the design and require the consumers to not store these events references once receaved).

I would be more worried about tons of logger.Debug we have instead :)

Reply:
hmm ok this is not too bad, AsyncProvider is becoming more then a background worker though, here the general component will be more useful.
Injecting Signals is Too much work? I am fine if it stays like this for now.

Reply:
Yeah too many classes to touch just to refactor, will cause this PR to contains even more changes, I'd change this once we chose to go (or not to go) with the ICorecomponent thing